File: win-prog/ProgWinVB/Paths Regions and Clipping/KeyholeClipCentered/KeyholeClipCentered.vb
'----------------------------------------------------
' KeyholeClipCentered.vb (c) 2002 by Charles Petzold
'----------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class KeyholeClipCentered
    Inherits KeyholeClip

    Shared Shadows Sub Main()
        Application.Run(new KeyholeClipCentered())
    End Sub

    Sub New()
        Text &= " Centered"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        grfx.SetClip(path)

        Dim rectf As RectangleF = path.GetBounds()
        Dim xOffset As Integer = CInt((cx - rectf.Width) / 2 - rectf.X)
        Dim yOffset As Integer = CInt((cy - rectf.Height) / 2 - rectf.Y)

        grfx.TranslateClip(xOffset, yOffset)
        grfx.DrawImage(img, xOffset, yOffset, img.Width, img.Height)
    End Sub
End Class

File: win-prog/ProgWinCS/Paths Regions and Clipping/KeyholeClipCentered/KeyholeClipCentered.cs
//--------------------------------------------------
// KeyholeClipCentered.cs © 2001 by Charles Petzold
//--------------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class KeyholeClipCentered: KeyholeClip
{
     public new static void Main()
     {
          Application.Run(new KeyholeClipCentered());
     }
     public KeyholeClipCentered()
     {
          Text += " Centered";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          grfx.SetClip(path);

          RectangleF rectf   = path.GetBounds();
          int        xOffset = (int)((cx - rectf.Width)  / 2 - rectf.X);
          int        yOffset = (int)((cy - rectf.Height) / 2 - rectf.Y);

          grfx.TranslateClip(xOffset, yOffset);
          grfx.DrawImage(image, xOffset, yOffset, image.Width, image.Height);
     }
}
